cli_tools/common/utils/logging/logger.go

```diff
@@ -15,50 +15,205 @@
+	log.SetOutput(redirectShim{logWriter})
+}
+
+// redirectShim forwards Go's standard logger to LogWriter.User.
+type redirectShim struct {
+	writer LogWriter
+}
+
+func (l redirectShim) Write(p []byte) (n int, err error) {
+	l.writer.WriteUser(string(p))
+	return len(p), nil
+}
+
+// LogReader exposes pb.OutputInfo to a consumer.
+type LogReader interface {
+	ReadOutputInfo() *pb.OutputInfo
+}
+
+// ToolLogger is a logger for interactive tools. It supports
+// string messages and structured metrics.
+//
+// Structured metrics are accumulated over the lifespan of the logger.
+type ToolLogger interface {
+	LogWriter
+	LogReader
 }
 
-func newLogEntry(message string) *LogEntry {
-	return &LogEntry{LocalTimestamp: time.Now(), Message: message}
+// defaultToolLogger is an implementation of ToolLogger that writes to an arbitrary writer.
+// It has the following behavior for each level:
+//
+// User:
+//  - Writes to the underlying log.Logger with an optional prefix. The prefix is used by
+//    gcloud and the web console for filtering which logs are shown to the user.
+//  - In addition to writing to the underlying log.Logger, the messages are buffered for
+//    inclusion in OutputInfo.SerialOutputs.
+// Debug:
+//  - Writes to the underlying log.Logger with an optional prefix.
+//  - In addition to writing to the underlying log.Logger, the messages are buffered for
+//    inclusion in OutputInfo.SerialOutputs.
+// Trace:
+//  - Included in OutputInfo.SerialOutputs
+type defaultToolLogger struct {
+	// userPrefix and debugPrefix are strings that are prepended to user and debug messages.
+	// The userPrefix string should be kept in sync with the matcher used by gcloud and the
+	// web UI when determining which log messages to show to the user.
+	userPrefix, debugPrefix string
+
+	// output: Destination for user and debug messages.
+	output *log.Logger
+
+	// trace: Buffer of trace messages. Cleared when ReadOutputInfo is called.
+	trace []string
+
+	// stdoutLogger: Buffer of messages sent to the output logger. Cleared when
+	// ReadOutputInfo is called.
+	userAndDebugBuffer strings.Builder
+
+	// outputInfo: View of OutputInfo that is updated when WriteMetric is called.
+	// Reset when ReadOutputInfo info is called.
+	outputInfo *pb.OutputInfo
+
+	// timestampFormat is the format string used when writing the current time in user and debug messages.
+	timestampFormat string
+
+	// timeProvider is a function that returns the current time. Typically time.Now. Exposed for testing.
+	timeProvider func() time.Time
+
+	// mutationLock should be taken when reading or writing trace, userAndDebugBuffer, or outputInfo.
+	mutationLock sync.Mutex
 }
 
-func (e *LogEntry) String() string {
-	return fmt.Sprintf("%s %s", e.LocalTimestamp.Format("2006-01-02T15:04:05Z"), e.Message)
+// WriteUser writes message to the underlying log.Logger, and then buffers the message
+// for inclusion in ReadOutputInfo().
+func (l *defaultToolLogger) WriteUser(message string) {
```

Comment:
Have you considered using "Info" instead of "User" for this category of messages? This would be more in line with existing logging framework's categorizations (e.g. INFO, WARNING, DEBUG, TRACE). We can explicitly state INFO is shown to users. This way, we can also use these messages for other purposes except for showing them to users.

Renaming it to "Info" would also avoid confusion on why `WriteUser` is not writing a user(name) but is actually logging.

Reply:
Resolving in favor of the "From the context of logging" thread